test/framework/src/main/java/org/elasticsearch/test/RandomObjects.java

```diff
@@ -216,39 +220,96 @@ private static Object randomFieldValue(Random random, int dataType) {
+     * method like {@link ShardInfo#fromXContent(XContentParser)}. A `withShardFailures` parameter indicates if
+     * the randomized ShardInfo must or must not contain shard failures.
      *
-     * @param random   Random generator
-     * @param failures If true, the {@link ShardInfo} will have random failures
-     * @return a random {@link ShardInfo}
+     * @param random            Random generator
+     * @param withShardFailures indicates if the generated ShardInfo must contain shard failures
      */
-    public static ShardInfo randomShardInfo(Random random, boolean failures) {
+    public static Tuple<ShardInfo, ShardInfo> randomShardInfo(Random random, boolean withShardFailures) {
         int total = randomIntBetween(random, 1, 10);
-        if (failures == false) {
-            return new ShardInfo(total, total);
+        if (withShardFailures == false) {
+            return Tuple.tuple(new ShardInfo(total, total), new ShardInfo(total, total));
         }
 
-        int successful = randomIntBetween(random, 1, total);
-        return new ShardInfo(total, successful, randomShardInfoFailures(random, Math.max(1, (total - successful))));
-    }
+        int successful = randomIntBetween(random, 1, Math.max(1, (total - 1)));
+        int failures = Math.max(1, (total - successful));
+
+        Failure[] actualFailures = new Failure[failures];
+        Failure[] expectedFailures = new Failure[failures];
 
-    public static ShardInfo.Failure[] randomShardInfoFailures(Random random, int nbFailures) {
-        List<ShardInfo.Failure> randomFailures = new ArrayList<>(nbFailures);
-        for (int i = 0; i < nbFailures; i++) {
-            randomFailures.add(randomShardInfoFailure(random));
+        for (int i = 0; i < failures; i++) {
+            Tuple<Failure, Failure> failure = randomShardInfoFailure(random);
+            actualFailures[i] = failure.v1();
+            expectedFailures[i] = failure.v2();
         }
-        return randomFailures.toArray(new ShardInfo.Failure[nbFailures]);
+        return Tuple.tuple(new ShardInfo(total, successful, actualFailures), new ShardInfo(total, successful, expectedFailures));
     }
 
-    public static ShardInfo.Failure randomShardInfoFailure(Random random) {
+    /**
+     * Returns a tuple that contains a randomized {@link Failure} value (left side) and its corresponding
+     * value (right side) after it has been printed out as a {@link ToXContent} and parsed back using a parsing
+     * method like {@link ShardInfo.Failure#fromXContent(XContentParser)}.
+     *
+     * @param random Random generator
+     */
+    private static Tuple<Failure, Failure> randomShardInfoFailure(Random random) {
         String index = randomAsciiOfLength(random, 5);
         String indexUuid = randomAsciiOfLength(random, 5);
         int shardId = randomIntBetween(random, 1, 10);
+        String nodeId = randomAsciiOfLength(random, 5);
+        RestStatus status = randomFrom(random, RestStatus.values());
```

Comment:
Is it okay that we might use a good Rest Status code in the Failure (e.g. 200)? I guess it doesn't matter or should we exclude those?